Decoding 10-bit source with 8-bit decoder?

### Problem Description

When using an 8-bit encoder to encode a 10-bit source (software avcodec decoder), the log shows the 8-bit decoder is used:

[19:57:46]  * video track
[19:57:46]    + decoder: hevc 8-bit

What is happening here?

(1) the log is just reading the wrong field (job->pix_fmt instead of pv->context->pix_fmt), and we actually decode 10-bit until reaching the encoder (or the first filter requiring 8-bit)

(2) because we're using an 8-bit encoder, HandBrake is configuring the avcodec decoder wrapper to convert the decoded output to 8-bit from the start (and everything is done in 8-bit throughout)



### Activity Log, Crash Log or any other details

```text
[19:55:05] macgui: Handbrake Version: 1.4.1 (2021081500)
[19:55:05] Compile-time hardening features are enabled
[19:55:06] hb_init: starting libhb thread
[19:57:07] macgui: ScanCore scanning titles with a duration of 100 seconds or more
[19:57:07] CPU: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
[19:57:07]  - Intel microarchitecture Kaby Lake
[19:57:07]  - logical processor count: 12
[19:57:07] hb_scan: path=/Users/tim/Desktop/_plexxx/_more-samples/HDR10-M2TS/00024.m2ts, title_index=0
udfread ERROR: ECMA 167 Volume Recognition failed
disc.c:333: failed opening UDF image /Users/tim/Desktop/_plexxx/_more-samples/HDR10-M2TS/00024.m2ts
disc.c:437: error opening file BDMV/index.bdmv
disc.c:437: error opening file BDMV/BACKUP/index.bdmv
bluray.c:2646: nav_get_title_list(/Users/tim/Desktop/_plexxx/_more-samples/HDR10-M2TS/00024.m2ts/) failed
[19:57:07] bd: not a bd - trying as a stream/file instead
libdvdread: DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.IFO failed
libdvdnav: vm: vm: failed to read VIDEO_TS.IFO
[19:57:07] dvd: not a dvd - trying as a stream/file instead
[19:57:07] file is MPEG Transport Stream with 192 byte packets offset 4 bytes
[19:57:07] hb_ts_stream_find_pids - end of file
Input #0, mpegts, from '/Users/tim/Desktop/_plexxx/_more-samples/HDR10-M2TS/00024.m2ts':
  Duration: 00:00:24.03, start: 4199.000000, bitrate: 47689 kb/s
  Program 1 
  Stream #0:0[0x1011]: Video: hevc (Main 10) (HDMV / 0x564D4448), yuv420p10le(tv, bt2020nc/bt2020/smpte2084), 3840x2160 [SAR 1:1 DAR 16:9], 23.98 fps, 23.98 tbr, 90k tbn, 23.98 tbc
  Stream #0:1[0x1100]: Audio: ac3 (AC-3 / 0x332D4341), 48000 Hz, 5.1(side), fltp, 640 kb/s
[19:57:07] scan: decoding previews for title 1
[19:57:07] file is MPEG Transport Stream with 192 byte packets offset 4 bytes
[19:57:07] hb_ts_stream_find_pids - end of file
[19:57:07] scan: audio 0x1: ac3, rate=48000Hz, bitrate=640000 Unknown (AC3) (5.1 ch) (640 kbps)
[19:57:18] scan: 20 previews, 3840x2160, 23.976 fps, autocrop = 0/0/0/0, aspect 16:9, PAR 1:1, color profile: 9-16-9
[19:57:18] scan: mastering display metadata: r(0.6800,0.3200) g(0.2650,0.6900) b(0.1500 0.0600) wp(0.3127, 0.3290) min_luminance=0.005000, max_luminance=1000.000000
[19:57:18] libhb: scan thread found 1 valid title(s)
[19:57:18] macgui: ScanCore scan done
[19:57:40] Compile-time hardening features are enabled
[19:57:40] hb_init: starting libhb thread
[19:57:40] macgui: fr.handbrake.HandBrakeXPCService scanning specifically for title: 1
[19:57:40] CPU: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
[19:57:40]  - Intel microarchitecture Kaby Lake
[19:57:40]  - logical processor count: 12
[19:57:40] hb_scan: path=/Users/tim/Desktop/_plexxx/_more-samples/HDR10-M2TS/00024.m2ts, title_index=1
udfread ERROR: ECMA 167 Volume Recognition failed
disc.c:333: failed opening UDF image /Users/tim/Desktop/_plexxx/_more-samples/HDR10-M2TS/00024.m2ts
disc.c:437: error opening file BDMV/index.bdmv
disc.c:437: error opening file BDMV/BACKUP/index.bdmv
bluray.c:2646: nav_get_title_list(/Users/tim/Desktop/_plexxx/_more-samples/HDR10-M2TS/00024.m2ts/) failed
[19:57:40] bd: not a bd - trying as a stream/file instead
libdvdread: DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.IFO failed
libdvdnav: vm: vm: failed to read VIDEO_TS.IFO
[19:57:40] dvd: not a dvd - trying as a stream/file instead
[19:57:40] file is MPEG Transport Stream with 192 byte packets offset 4 bytes
[19:57:40] hb_ts_stream_find_pids - end of file
Input #0, mpegts, from '/Users/tim/Desktop/_plexxx/_more-samples/HDR10-M2TS/00024.m2ts':
  Duration: 00:00:24.03, start: 4199.000000, bitrate: 47689 kb/s
  Program 1 
  Stream #0:0[0x1011]: Video: hevc (Main 10) (HDMV / 0x564D4448), yuv420p10le(tv, bt2020nc/bt2020/smpte2084), 3840x2160 [SAR 1:1 DAR 16:9], 23.98 fps, 23.98 tbr, 90k tbn, 23.98 tbc
  Stream #0:1[0x1100]: Audio: ac3 (AC-3 / 0x332D4341), 48000 Hz, 5.1(side), fltp, 640 kb/s
[19:57:40] scan: decoding previews for title 1
[19:57:40] file is MPEG Transport Stream with 192 byte packets offset 4 bytes
[19:57:40] hb_ts_stream_find_pids - end of file
[19:57:40] scan: audio 0x1: ac3, rate=48000Hz, bitrate=640000 Unknown (AC3) (5.1 ch) (640 kbps)
[19:57:46] scan: 10 previews, 3840x2160, 23.976 fps, autocrop = 0/0/0/0, aspect 16:9, PAR 1:1, color profile: 9-16-9
[19:57:46] scan: mastering display metadata: r(0.6800,0.3200) g(0.2650,0.6900) b(0.1500 0.0600) wp(0.3127, 0.3290) min_luminance=0.005000, max_luminance=1000.000000
[19:57:46] libhb: scan thread found 1 valid title(s)
[19:57:46] macgui: fr.handbrake.HandBrakeXPCService scan done
[19:57:46] macgui: fr.handbrake.HandBrakeXPCService started encoding 00024 -- rf20.00 -- 1.mkv
[19:57:46] macgui: fr.handbrake.HandBrakeXPCService with preset HDR2SDR#1 (Modified)
[19:57:46] Starting work at: Wed Aug 25 19:57:46 2021

[19:57:46] 1 job(s) to process
[19:57:46] Starting Task: Encoding Pass
[19:57:46] Skipping crop/scale filter
[19:57:46] job configuration:
[19:57:46]  * source
[19:57:46]    + /Users/tim/Desktop/_plexxx/_more-samples/HDR10-M2TS/00024.m2ts
[19:57:46]    + title 1, chapter(s) 1 to 1
[19:57:46]    + container: mpegts
[19:57:46]    + data rate: 47689 kbps
[19:57:46]  * destination
[19:57:46]    + /Users/tim/Downloads/00024 -- rf20.00 -- 1.mkv
[19:57:46]    + container: Matroska (libavformat)
[19:57:46]  * video track
[19:57:46]    + decoder: hevc 8-bit
[19:57:46]    + filters
[19:57:46]      + Framerate Shaper (mode=1:rate=27000000/1126125)
[19:57:46]        + frame rate: 23.976 fps -> constant 23.976 fps
[19:57:46]      + Colorspace (primaries=bt709:transfer=bt709:matrix=bt709:tonemap=hable:desat=0)
[19:57:46]    + Output geometry
[19:57:46]      + storage dimensions: 3840 x 2160
[19:57:46]      + pixel aspect ratio: 1 : 1
[19:57:46]      + display dimensions: 3840 x 2160
[19:57:46]    + encoder: H.265 (libx265)
[19:57:46]      + preset:  medium
[19:57:46]      + profile: auto
[19:57:46]      + level:   auto
[19:57:46]      + quality: 20.00 (RF)
[19:57:46]      + color profile: 1-1-1
[19:57:46] file is MPEG Transport Stream with 192 byte packets offset 4 bytes
[19:57:46] hb_ts_stream_find_pids - end of file
[19:57:46] sync: expecting 576 video frames
x265 [info]: HEVC encoder version 3.5+1-f0c1022b6
x265 [info]: build info [Mac OS X][clang 12.0.5][64 bit] 8bit+10bit+12bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [info]: Main profile, Level-5 (Main tier)
x265 [info]: Thread pool created using 12 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(34 rows)
x265 [info]: Coding QT: max CU size, min CU size : 64 / 8
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : hex / 57 / 2 / 3
x265 [info]: Keyframe min / max / scenecut / bias  : 24 / 240 / 40 / 5.00 
x265 [info]: Lookahead / bframes / badapt        : 20 / 4 / 2
x265 [info]: b-pyramid / weightp / weightb       : 1 / 1 / 0
x265 [info]: References / ref-limit  cu / depth  : 3 / off / on
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 2 / 1.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-20.0 / 0.60
x265 [info]: tools: rd=3 psy-rd=2.00 early-skip rskip mode=1 signhide tmvp
x265 [info]: tools: b-intra strong-intra-smoothing lslices=8 deblock sao
[19:57:47] sync: first pts video is 0
[19:57:47] sync: Chapter 1 at frame 1 time 0
[19:59:42] reader: done. 1 scr changes
[19:59:56] work: average encoding speed for job is 4.575088 fps
[19:59:57] vfr: 576 frames output, 0 dropped and 0 duped for CFR/PFR
[19:59:57] vfr: lost time: 0 (0 frames)
[19:59:57] vfr: gained time: 0 (0 frames) (0 not accounted for)
[19:59:57] hevc-decoder done: 576 frames, 0 decoder errors
[19:59:57] sync: got 576 frames, 576 expected
[19:59:57] sync: framerate min 23.976 fps, max 23.976 fps, avg 23.976 fps
x265 [info]: frame I:      3, Avg QP:18.42  kb/s: 48115.02
x265 [info]: frame P:    191, Avg QP:20.74  kb/s: 21748.18
x265 [info]: frame B:    382, Avg QP:24.34  kb/s: 4727.73 
x265 [info]: Weighted P-Frames: Y:4.7% UV:4.7%
x265 [info]: consecutive B-frames: 12.4% 19.6% 36.1% 22.7% 9.3% 

encoded 576 frames in 130.76s (4.40 fps), 10597.64 kb/s, Avg QP:23.11
[19:59:57] mux: track 0, 576 frames, 31827007 bytes, 10598.40 kbps, fifo 256
[19:59:57] Finished work at: Wed Aug 25 19:59:57 2021

[19:59:57] libhb: work result = 0
[19:59:57] macgui: Queue Done, there are no more pending encodes
[19:59:57] macgui: fr.handbrake.HandBrakeXPCService work done
```


### What Operating System are you running?

macOS 10.15 Catalina

### What version of HandBrake are you running?

HandBrake.app 1.4.1

### Where did you download HandBrake from?

handbrake.fr

--- a/libhb/common.c
+++ b/libhb/common.c
@@ -5934,25 +5934,12 @@ static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
 
     if (job->title->video_decode_support & HB_DECODE_SUPPORT_QSV)
     {
-#if HB_PROJECT_FEATURE_QSV && (defined( _WIN32 ) || defined( __MINGW32__ ))
-        if (hb_qsv_full_path_is_enabled(job))
+        // Allow formats supported by QSV pipeline via system memory
+        // at that stage only, video memory formats will be reassigned later if allowed
+        if (pix_fmt != AV_PIX_FMT_YUV420P10 &&
+            pix_fmt != AV_PIX_FMT_YUV420P)
         {
-            // Formats supported in QSV pipeline via video memory
-            if (pix_fmt != AV_PIX_FMT_P010LE &&
-                pix_fmt != AV_PIX_FMT_NV12)
-            {
-                return 0;
-            }
-        }
-        else
-#endif
-        {
-            // Formats supported by QSV pipeline via system memory
-            if (pix_fmt != AV_PIX_FMT_YUV420P10 &&
-                pix_fmt != AV_PIX_FMT_YUV420P)
-            {
-                return 0;
-            }
+            return 0;
         }
     }
     else

--- a/libhb/common.c
+++ b/libhb/common.c
@@ -5934,7 +5934,7 @@ static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
 
     if (job->title->video_decode_support & HB_DECODE_SUPPORT_QSV)
     {
-#if defined(_WIN32) || defined(__MINGW32__)
+#if HB_PROJECT_FEATURE_QSV && (defined( _WIN32 ) || defined( __MINGW32__ ))
         if (hb_qsv_full_path_is_enabled(job))
         {
             // Formats supported in QSV pipeline via video memory

--- a/libhb/common.c
+++ b/libhb/common.c
@@ -5955,6 +5955,16 @@ static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
             }
         }
     }
+    else
+    {
+        // Allow biplanar formats only if
+        // hardware decoding is enabled.
+        if (pix_fmt == AV_PIX_FMT_P010LE ||
+            pix_fmt == AV_PIX_FMT_NV12)
+        {
+            return 0;
+        }
+    }
 
     for (int i = 0; i < hb_list_count(job->list_filter); i++)
     {

--- a/libhb/common.c
+++ b/libhb/common.c
@@ -1678,6 +1678,50 @@ const char* const* hb_video_encoder_get_levels(int encoder)
     }
 }
 
+static const enum AVPixelFormat standard_pix_fmts[] =
+{
+    AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
+static const enum AVPixelFormat standard_10bit_pix_fmts[] =
+{
+    AV_PIX_FMT_YUV420P10, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
+static const enum AVPixelFormat standard_12bit_pix_fmts[] =
+{
+    AV_PIX_FMT_YUV420P12, AV_PIX_FMT_YUV420P10, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
+const int* hb_video_encoder_get_pix_fmts(int encoder)
+{
+#if HB_PROJECT_FEATURE_QSV
+    if (encoder & HB_VCODEC_QSV_MASK)
+    {
+        return hb_qsv_get_pix_fmts(encoder);
+    }
+#endif
+
+    if (encoder & HB_VCODEC_FFMPEG_MASK)
+    {
+        return hb_av_get_pix_fmts(encoder);
+    }
+
+    switch (encoder)
+    {
+        case HB_VCODEC_X264_10BIT:
+            return standard_10bit_pix_fmts;
+#if HB_PROJECT_FEATURE_X265
+        case HB_VCODEC_X265_10BIT:
+            return standard_10bit_pix_fmts;
+        case HB_VCODEC_X265_12BIT:
+            return standard_12bit_pix_fmts;
+#endif
+        default:
+            return standard_pix_fmts;
+    }
+}
+
 // Get limits and hints for the UIs.
 //
 // granularity sets the minimum step increments that should be used
@@ -5847,7 +5891,7 @@ int hb_get_bit_depth(int format)
     int i, min, max;
 
     if (!desc || !desc->nb_components) {
-        min = max = 0;
+        return -1;
     }
 
     min = INT_MAX, max = -INT_MAX;
@@ -5859,36 +5903,27 @@ int hb_get_bit_depth(int format)
     return max;
 }
 
-static int bit_depth_is_supported(hb_job_t * job, int bit_depth)
+static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
 {
-    for (int i = 0; i < hb_list_count(job->list_filter); i++)
-    {
-        hb_filter_object_t *filter = hb_list_item(job->list_filter, i);
+    int title_bit_depth = hb_get_bit_depth(job->title->pix_fmt);
+    int pix_fmt_bit_depth = hb_get_bit_depth(pix_fmt);
 
-        switch (filter->id) {
-            case HB_FILTER_DETELECINE:
-            case HB_FILTER_COMB_DETECT:
-            case HB_FILTER_DECOMB:
-            case HB_FILTER_DENOISE:
-            case HB_FILTER_NLMEANS:
-            case HB_FILTER_CHROMA_SMOOTH:
-            case HB_FILTER_LAPSHARP:
-            case HB_FILTER_UNSHARP:
-            case HB_FILTER_GRAYSCALE:
-                return 0;
-        }
-    }
-
-    if (hb_video_encoder_get_depth(job->vcodec) < bit_depth)
+    if (pix_fmt_bit_depth > title_bit_depth)
     {
         return 0;
     }
 
-    return 1;
-}
+    // Allow biplanar formats only if
+    // hardware decoder is supported.
+    if (pix_fmt == AV_PIX_FMT_P010LE ||
+        pix_fmt == AV_PIX_FMT_NV12)
+    {
+        if (!(job->title->video_decode_support & HB_DECODE_SUPPORT_QSV))
+        {
+            return 0;
+        }
+    }
 
-static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
-{
     for (int i = 0; i < hb_list_count(job->list_filter); i++)
     {
         hb_filter_object_t *filter = hb_list_item(job->list_filter, i);
@@ -5908,6 +5943,13 @@ static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
                 {
                     return 0;
                 }
+            case HB_FILTER_RENDER_SUB:
+               if (pix_fmt != AV_PIX_FMT_YUV420P   ||
+                   pix_fmt != AV_PIX_FMT_YUV420P10 ||
+                   pix_fmt != AV_PIX_FMT_YUV420P12)
+               {
+                   return 0;
+               }
         }
     }
 
@@ -5916,42 +5958,16 @@ static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
 
 int hb_get_best_pix_fmt(hb_job_t * job)
 {
-    int bit_depth = hb_get_bit_depth(job->title->pix_fmt);
-#if HB_PROJECT_FEATURE_QSV && (defined( _WIN32 ) || defined( __MINGW32__ ))
-    if (hb_qsv_encoder_info_get(hb_qsv_get_adapter_index(), job->vcodec))
+    const int *pix_fmts = hb_video_encoder_get_pix_fmts(job->vcodec);
+
+    while (*pix_fmts != AV_PIX_FMT_NONE)
     {
-        if (hb_qsv_full_path_is_enabled(job))
+        if (pix_fmt_is_supported(job, *pix_fmts))
         {
-            if (job->title->pix_fmt == AV_PIX_FMT_YUV420P10 && job->vcodec == HB_VCODEC_QSV_H265_10BIT)
-            {
-                return AV_PIX_FMT_P010LE;
-            }
-            else
-            {
-                return AV_PIX_FMT_NV12;
-            }
+            return *pix_fmts;
         }
-        else
-        {
-            // system memory usage: QSV encoder only or QSV decoder + SW filters + QSV encoder
-            return AV_PIX_FMT_YUV420P;
-        }
+        pix_fmts++;
     }
-#endif
-    if (job->vcodec == HB_VCODEC_FFMPEG_MF_H264 || job->vcodec == HB_VCODEC_FFMPEG_MF_H265)
-    {
-        if (pix_fmt_is_supported(job, AV_PIX_FMT_NV12))
-        {
-            return AV_PIX_FMT_NV12;
-        }
-    }
-    if (bit_depth >= 12 && bit_depth_is_supported(job, 12))
-    {
-        return AV_PIX_FMT_YUV420P12;
-    }
-    if (bit_depth >= 10 && bit_depth_is_supported(job, 10))
-    {
-        return AV_PIX_FMT_YUV420P10;
-    }
+
     return AV_PIX_FMT_YUV420P;
 }

--- a/libhb/common.c
+++ b/libhb/common.c
@@ -5932,14 +5932,27 @@ static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
         return 0;
     }
 
-    // Allow biplanar formats only if
-    // hardware decoding is enabled.
-    if (pix_fmt == AV_PIX_FMT_P010LE ||
-        pix_fmt == AV_PIX_FMT_NV12)
+    if (job->title->video_decode_support & HB_DECODE_SUPPORT_QSV)
     {
-        if (!(job->title->video_decode_support & HB_DECODE_SUPPORT_QSV))
+#if defined(_WIN32) || defined(__MINGW32__)
+        if (hb_qsv_full_path_is_enabled(job))
         {
-            return 0;
+            // Formats supported in QSV pipeline via video memory
+            if (pix_fmt != AV_PIX_FMT_P010LE &&
+                pix_fmt != AV_PIX_FMT_NV12)
+            {
+                return 0;
+            }
+        }
+        else
+#endif
+        {
+            // Formats supported by QSV pipeline via system memory
+            if (pix_fmt != AV_PIX_FMT_YUV420P10 &&
+                pix_fmt != AV_PIX_FMT_YUV420P)
+            {
+                return 0;
+            }
         }
     }
 

--- a/libhb/common.c
+++ b/libhb/common.c
@@ -1516,6 +1516,20 @@ int hb_video_encoder_is_supported(int encoder)
     return 0;
 }
 
+int hb_video_encoder_pix_fmt_is_supported(int encoder, int pix_fmt)
+{
+    const int *pix_fmts = hb_video_encoder_get_pix_fmts(encoder);
+    while (*pix_fmts != AV_PIX_FMT_NONE)
+    {
+        if (pix_fmt == *pix_fmts)
+        {
+            return 1;
+        }
+        pix_fmts++;
+    }
+    return 0;
+}
+
 int hb_video_encoder_get_depth(int encoder)
 {
     switch (encoder)
@@ -3921,7 +3935,8 @@ static void job_setup(hb_job_t * job, hb_title_t * title)
     job->pass_id    = HB_PASS_ENCODE;
     job->vrate      = title->vrate;
 
-    job->pix_fmt        = AV_PIX_FMT_YUV420P;
+    job->input_pix_fmt  = AV_PIX_FMT_YUV420P;
+    job->output_pix_fmt = AV_PIX_FMT_YUV420P;
     job->color_prim     = title->color_prim;
     job->color_transfer = title->color_transfer;
     job->color_matrix   = title->color_matrix;
@@ -4330,6 +4345,10 @@ hb_filter_object_t * hb_filter_get( int filter_id )
 
         case HB_FILTER_GRAYSCALE:
             filter = &hb_filter_grayscale;
+            break;
+
+        case HB_FILTER_FORMAT:
+            filter = &hb_filter_format;
             break;
 
 #if HB_PROJECT_FEATURE_QSV
@@ -5944,8 +5963,8 @@ static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
                     return 0;
                 }
             case HB_FILTER_RENDER_SUB:
-               if (pix_fmt != AV_PIX_FMT_YUV420P   ||
-                   pix_fmt != AV_PIX_FMT_YUV420P10 ||
+               if (pix_fmt != AV_PIX_FMT_YUV420P   &&
+                   pix_fmt != AV_PIX_FMT_YUV420P10 &&
                    pix_fmt != AV_PIX_FMT_YUV420P12)
                {
                    return 0;
@@ -5956,9 +5975,14 @@ static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
     return 1;
 }
 
+static const enum AVPixelFormat pipeline_pix_fmts[] =
+{
+    AV_PIX_FMT_YUV420P12, AV_PIX_FMT_P010LE, AV_PIX_FMT_YUV420P10, AV_PIX_FMT_NV12, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
 int hb_get_best_pix_fmt(hb_job_t * job)
 {
-    const int *pix_fmts = hb_video_encoder_get_pix_fmts(job->vcodec);
+    const int *pix_fmts = pipeline_pix_fmts;
 
     while (*pix_fmts != AV_PIX_FMT_NONE)
     {

--- a/libhb/common.c
+++ b/libhb/common.c
@@ -5933,7 +5933,7 @@ static int pix_fmt_is_supported(hb_job_t * job, int pix_fmt)
     }
 
     // Allow biplanar formats only if
-    // hardware decoder is supported.
+    // hardware decoding is enabled.
     if (pix_fmt == AV_PIX_FMT_P010LE ||
         pix_fmt == AV_PIX_FMT_NV12)
     {

--- a/libhb/decavcodec.c
+++ b/libhb/decavcodec.c
@@ -1174,7 +1174,7 @@ int reinit_video_filters(hb_work_private_t * pv)
             orig_width = pv->job->title->geometry.width;
             orig_height = pv->job->title->geometry.height;
         }
-        pix_fmt = pv->job->pix_fmt;
+        pix_fmt = pv->job->input_pix_fmt;
         color_range = pv->job->color_range;
     }
 

--- a/libhb/enc_qsv.c
+++ b/libhb/enc_qsv.c
@@ -1,3 +1,4 @@
+
 /* ********************************************************************* *\
 
 Copyright (C) 2013 Intel Corporation.  All rights reserved.
@@ -850,7 +851,7 @@ int qsv_enc_init(hb_work_private_t *pv)
         {
             pv->sws_context_to_nv12 = hb_sws_get_context(
                                         job->width, job->height,
-                                        AV_PIX_FMT_YUV420P, job->color_range,
+                                        job->pix_fmt, job->color_range,
                                         job->width, job->height,
                                         AV_PIX_FMT_P010LE, job->color_range,
                                         SWS_LANCZOS|SWS_ACCURATE_RND,
@@ -860,7 +861,7 @@ int qsv_enc_init(hb_work_private_t *pv)
         {
             pv->sws_context_to_nv12 = hb_sws_get_context(
                                         job->width, job->height,
-                                        AV_PIX_FMT_YUV420P, job->color_range,
+                                        job->pix_fmt, job->color_range,
                                         job->width, job->height,
                                         AV_PIX_FMT_NV12, job->color_range,
                                         SWS_LANCZOS|SWS_ACCURATE_RND,

--- a/libhb/enc_qsv.c
+++ b/libhb/enc_qsv.c
@@ -851,7 +851,7 @@ int qsv_enc_init(hb_work_private_t *pv)
         {
             pv->sws_context_to_nv12 = hb_sws_get_context(
                                         job->width, job->height,
-                                        job->pix_fmt, job->color_range,
+                                        job->output_pix_fmt, job->color_range,
                                         job->width, job->height,
                                         AV_PIX_FMT_P010LE, job->color_range,
                                         SWS_LANCZOS|SWS_ACCURATE_RND,
@@ -861,7 +861,7 @@ int qsv_enc_init(hb_work_private_t *pv)
         {
             pv->sws_context_to_nv12 = hb_sws_get_context(
                                         job->width, job->height,
-                                        job->pix_fmt, job->color_range,
+                                        job->output_pix_fmt, job->color_range,
                                         job->width, job->height,
                                         AV_PIX_FMT_NV12, job->color_range,
                                         SWS_LANCZOS|SWS_ACCURATE_RND,

--- a/libhb/encavcodec.c
+++ b/libhb/encavcodec.c
@@ -928,7 +928,6 @@ int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
     {
         av_dict_set(&av_opts, "profile", "main10", 0);
         context->max_b_frames = 16;
-        context->pix_fmt = AV_PIX_FMT_P010LE;
     }
 
     if (job->vcodec == HB_VCODEC_FFMPEG_VCE_H264)
@@ -1035,7 +1034,6 @@ int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
         job->vcodec == HB_VCODEC_FFMPEG_MF_H265)
     {
         av_dict_set(&av_opts, "hw_encoding", "1", 0);
-        context->pix_fmt = AV_PIX_FMT_NV12;
     }
 
     if (job->vcodec == HB_VCODEC_FFMPEG_MF_H265)

--- a/libhb/encavcodec.c
+++ b/libhb/encavcodec.c
@@ -117,6 +117,26 @@ static const char * const h265_mf_profile_name[] =
     "auto", "main",  NULL
 };
 
+static const enum AVPixelFormat pix_fmts[] =
+{
+    AV_PIX_FMT_NV12, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
+static const enum AVPixelFormat h26x_vt_pix_fmts[] =
+{
+    AV_PIX_FMT_NV12, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
+static const enum AVPixelFormat h265_10bit_vt_pix_fmts[] =
+{
+    AV_PIX_FMT_P010LE, AV_PIX_FMT_YUV420P10, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
+static const enum AVPixelFormat h26x_mf_pix_fmts[] =
+{
+    AV_PIX_FMT_NV12, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
 int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
 {
     int ret = 0;
@@ -911,22 +931,7 @@ int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
     {
         av_dict_set(&av_opts, "profile", "main10", 0);
         context->max_b_frames = 16;
-
-        if (job->pix_fmt != AV_PIX_FMT_P010LE)
-        {
-            pv->sws = hb_sws_get_context(
-                                         job->width, job->height,
-                                         context->pix_fmt, job->color_range,
-                                         job->width, job->height,
-                                         AV_PIX_FMT_P010LE, job->color_range,
-                                         SWS_LANCZOS|SWS_ACCURATE_RND,
-                                         SWS_CS_DEFAULT);
-
-            pv->sws_buf = hb_frame_buffer_init(
-                             AV_PIX_FMT_P010LE, job->width, job->height);
-
-            context->pix_fmt = AV_PIX_FMT_P010LE;
-        }
+        context->pix_fmt = AV_PIX_FMT_P010LE;
     }
 
     if (job->vcodec == HB_VCODEC_FFMPEG_VCE_H264)
@@ -1033,22 +1038,7 @@ int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
         job->vcodec == HB_VCODEC_FFMPEG_MF_H265)
     {
         av_dict_set(&av_opts, "hw_encoding", "1", 0);
-
-        if (job->pix_fmt != AV_PIX_FMT_NV12)
-        {
-            pv->sws = hb_sws_get_context(
-                                         job->width, job->height,
-                                         context->pix_fmt, job->color_range,
-                                         job->width, job->height,
-                                         AV_PIX_FMT_NV12, job->color_range,
-                                         SWS_LANCZOS|SWS_ACCURATE_RND,
-                                         SWS_CS_DEFAULT);
-
-            pv->sws_buf = hb_frame_buffer_init(
-                             AV_PIX_FMT_NV12, job->width, job->height);
-
-            context->pix_fmt = AV_PIX_FMT_NV12;
-        }
+        context->pix_fmt = AV_PIX_FMT_NV12;
     }
 
     if (job->vcodec == HB_VCODEC_FFMPEG_MF_H265)
@@ -1057,6 +1047,21 @@ int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
         // support setting this to either 1 or 2, while others only support
         // setting it to 1.
         context->max_b_frames = 1;
+    }
+
+    if (job->pix_fmt != context->pix_fmt)
+    {
+        // Some encoders require a specific input pixel format
+        // that could be different from the current pipeline format.
+        // Configure a sws context to handle this unlucky case.
+        pv->sws = hb_sws_get_context(
+                                     job->width, job->height,
+                                     job->pix_fmt, job->color_range,
+                                     job->width, job->height,
+                                     context->pix_fmt, job->color_range,
+                                     SWS_LANCZOS|SWS_ACCURATE_RND,
+                                     SWS_CS_DEFAULT);
+        pv->sws_buf = hb_frame_buffer_init(context->pix_fmt, job->width, job->height);
     }
 
     if( job->pass_id == HB_PASS_ENCODE_1ST ||
@@ -1632,5 +1637,23 @@ const char* const* hb_av_profile_get_names(int encoder)
 
          default:
              return NULL;
+     }
+}
+
+const int* hb_av_get_pix_fmts(int encoder)
+{
+    switch (encoder)
+    {
+        case HB_VCODEC_FFMPEG_VT_H264:
+        case HB_VCODEC_FFMPEG_VT_H265:
+            return h26x_vt_pix_fmts;
+        case HB_VCODEC_FFMPEG_VT_H265_10BIT:
+            return h265_10bit_vt_pix_fmts;
+        case HB_VCODEC_FFMPEG_MF_H264:
+        case HB_VCODEC_FFMPEG_MF_H265:
+            return h26x_mf_pix_fmts;
+
+         default:
+             return pix_fmts;
      }
 }

--- a/libhb/encavcodec.c
+++ b/libhb/encavcodec.c
@@ -51,9 +51,6 @@ struct hb_work_private_s
     } frame_info[FRAME_INFO_SIZE];
 
     hb_chapter_queue_t * chapter_queue;
-
-    struct SwsContext  * sws;
-    hb_buffer_t        * sws_buf;
 };
 
 int  encavcodecInit( hb_work_object_t *, hb_job_t * );
@@ -119,7 +116,7 @@ static const char * const h265_mf_profile_name[] =
 
 static const enum AVPixelFormat pix_fmts[] =
 {
-    AV_PIX_FMT_NV12, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+    AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
 };
 
 static const enum AVPixelFormat h26x_vt_pix_fmts[] =
@@ -129,12 +126,12 @@ static const enum AVPixelFormat h26x_vt_pix_fmts[] =
 
 static const enum AVPixelFormat h265_10bit_vt_pix_fmts[] =
 {
-    AV_PIX_FMT_P010LE, AV_PIX_FMT_YUV420P10, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+    AV_PIX_FMT_P010LE, AV_PIX_FMT_NONE
 };
 
 static const enum AVPixelFormat h26x_mf_pix_fmts[] =
 {
-    AV_PIX_FMT_NV12, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+    AV_PIX_FMT_NV12, AV_PIX_FMT_NONE
 };
 
 int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
@@ -862,7 +859,7 @@ int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
     }
     context->width     = job->width;
     context->height    = job->height;
-    context->pix_fmt   = job->pix_fmt;
+    context->pix_fmt   = job->output_pix_fmt;
 
     context->sample_aspect_ratio.num = job->par.num;
     context->sample_aspect_ratio.den = job->par.den;
@@ -1049,21 +1046,6 @@ int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
         context->max_b_frames = 1;
     }
 
-    if (job->pix_fmt != context->pix_fmt)
-    {
-        // Some encoders require a specific input pixel format
-        // that could be different from the current pipeline format.
-        // Configure a sws context to handle this unlucky case.
-        pv->sws = hb_sws_get_context(
-                                     job->width, job->height,
-                                     job->pix_fmt, job->color_range,
-                                     job->width, job->height,
-                                     context->pix_fmt, job->color_range,
-                                     SWS_LANCZOS|SWS_ACCURATE_RND,
-                                     SWS_CS_DEFAULT);
-        pv->sws_buf = hb_frame_buffer_init(context->pix_fmt, job->width, job->height);
-    }
-
     if( job->pass_id == HB_PASS_ENCODE_1ST ||
         job->pass_id == HB_PASS_ENCODE_2ND )
     {
@@ -1211,14 +1193,6 @@ void encavcodecClose( hb_work_object_t * w )
             avcodec_flush_buffers( pv->context );
         }
         hb_avcodec_free_context(&pv->context);
-    }
-    if (pv->sws != NULL)
-    {
-        sws_freeContext(pv->sws);
-    }
-    if (pv->sws_buf != NULL)
-    {
-        hb_buffer_close(&pv->sws_buf);
     }
     if( pv->file )
     {
@@ -1378,25 +1352,6 @@ static void Encode( hb_work_object_t *w, hb_buffer_t *in,
     frame.linesize[0] = in->plane[0].stride;
     frame.linesize[1] = in->plane[1].stride;
     frame.linesize[2] = in->plane[2].stride;
-
-    if (pv->sws)
-    {
-        uint8_t *srcs[]   = { in->plane[0].data, in->plane[1].data, in->plane[2].data };
-        int srcs_stride[] = { in->plane[0].stride, in->plane[1].stride, in->plane[2].stride };
-        uint8_t *dsts[]   = { pv->sws_buf->plane[0].data, pv->sws_buf->plane[1].data, NULL };
-        int dsts_stride[] = { pv->sws_buf->plane[0].stride, pv->sws_buf->plane[1].stride, 0 };
-
-        sws_scale(pv->sws,
-                  (const uint8_t* const*)srcs, srcs_stride,
-                  0, in->f.height, dsts, dsts_stride);
-
-        for (int i = 0; i < 3; i++)
-        {
-            frame.data[i] = dsts[i];
-            frame.linesize[i] = dsts_stride[i];
-        }
-        frame.format = pv->context->pix_fmt;
-    }
 
     if (in->s.new_chap > 0 && pv->job->chapter_markers)
     {

--- a/libhb/encx264.c
+++ b/libhb/encx264.c
@@ -756,7 +756,7 @@ static hb_buffer_t *x264_encode( hb_work_object_t *w, hb_buffer_t *in )
 
     /* Point x264 at our current buffers Y(UV) data.  */
     if (pv->pic_in.img.i_csp & X264_CSP_HIGH_DEPTH &&
-        job->pix_fmt == AV_PIX_FMT_YUV420P)
+        job->output_pix_fmt == AV_PIX_FMT_YUV420P)
     {
         tmp = expand_buf(pv->api->bit_depth, in);
         pv->pic_in.img.i_stride[0] = tmp->plane[0].stride;

--- a/libhb/encx265.c
+++ b/libhb/encx265.c
@@ -260,7 +260,7 @@ int encx265Init(hb_work_object_t *w, hb_job_t *job)
     }
 
     /* Bit depth */
-    pv->bit_depth = hb_get_bit_depth(job->pix_fmt);
+    pv->bit_depth = hb_get_bit_depth(job->output_pix_fmt);
 
     /* iterate through x265_opts and parse the options */
     hb_dict_t *x265_opts;

--- a/libhb/format.c
+++ b/libhb/format.c
@@ -1,0 +1,74 @@
+/* format.c
+
+   Copyright (c) 2003-2015 HandBrake Team
+   This file is part of the HandBrake source code
+   Homepage: <http://handbrake.fr/>.
+   It may be used under the terms of the GNU General Public License v2.
+   For full terms see the file COPYING file or visit http://www.gnu.org/licenses/gpl-2.0.html
+ */
+
+#include "handbrake/common.h"
+#include "handbrake/avfilter_priv.h"
+
+static int format_init(hb_filter_object_t * filter,
+                           hb_filter_init_t * init);
+
+const char format_template[] =
+    "format=^"HB_ALL_REG"$";
+
+hb_filter_object_t hb_filter_format =
+{
+    .id                = HB_FILTER_FORMAT,
+    .enforce_order     = 1,
+    .skip              = 1,
+    .name              = "Format",
+    .settings          = NULL,
+    .init              = format_init,
+    .work              = hb_avfilter_null_work,
+    .close             = hb_avfilter_alias_close,
+    .settings_template = format_template,
+};
+
+static int format_init(hb_filter_object_t * filter, hb_filter_init_t * init)
+{
+    hb_filter_private_t * pv = NULL;
+
+    pv = calloc(1, sizeof(struct hb_filter_private_s));
+    filter->private_data = pv;
+    if (pv == NULL)
+    {
+        return 1;
+    }
+    pv->input = *init;
+
+    hb_dict_t        * settings = filter->settings;
+
+    char * format = NULL;
+
+    hb_dict_extract_string(&format, settings, "format");
+
+    if (!format)
+    {
+        return 0;
+    }
+
+    hb_value_array_t * avfilters = hb_value_array_init();
+    hb_dict_t * avfilter   = NULL;
+    hb_dict_t * avsettings = NULL;
+
+    // Format
+    avfilter   = hb_dict_init();
+    avsettings = hb_dict_init();
+
+    hb_dict_set_string(avsettings, "pix_fmts", format);
+    hb_dict_set(avfilter, "format", avsettings);
+
+    hb_value_array_append(avfilters, avfilter);
+
+    pv->avfilters = avfilters;
+
+    init->pix_fmt = av_get_pix_fmt(format);
+    pv->output = *init;
+
+    return 0;
+}

--- a/libhb/format.c
+++ b/libhb/format.c
@@ -1,6 +1,6 @@
 /* format.c
 
-   Copyright (c) 2003-2015 HandBrake Team
+   Copyright (c) 2003-2021 HandBrake Team
    This file is part of the HandBrake source code
    Homepage: <http://handbrake.fr/>.
    It may be used under the terms of the GNU General Public License v2.
@@ -10,8 +10,8 @@
 #include "handbrake/common.h"
 #include "handbrake/avfilter_priv.h"
 
-static int format_init(hb_filter_object_t * filter,
-                           hb_filter_init_t * init);
+static int format_init(hb_filter_object_t *filter,
+                           hb_filter_init_t *init);
 
 const char format_template[] =
     "format=^"HB_ALL_REG"$";
@@ -29,9 +29,9 @@ hb_filter_object_t hb_filter_format =
     .settings_template = format_template,
 };
 
-static int format_init(hb_filter_object_t * filter, hb_filter_init_t * init)
+static int format_init(hb_filter_object_t *filter, hb_filter_init_t *init)
 {
-    hb_filter_private_t * pv = NULL;
+    hb_filter_private_t *pv = NULL;
 
     pv = calloc(1, sizeof(struct hb_filter_private_s));
     filter->private_data = pv;
@@ -41,24 +41,19 @@ static int format_init(hb_filter_object_t * filter, hb_filter_init_t * init)
     }
     pv->input = *init;
 
-    hb_dict_t        * settings = filter->settings;
-
-    char * format = NULL;
+    hb_dict_t *settings = filter->settings;
+    char *format = NULL;
 
     hb_dict_extract_string(&format, settings, "format");
 
-    if (!format)
+    if (format == NULL)
     {
         return 0;
     }
 
-    hb_value_array_t * avfilters = hb_value_array_init();
-    hb_dict_t * avfilter   = NULL;
-    hb_dict_t * avsettings = NULL;
-
-    // Format
-    avfilter   = hb_dict_init();
-    avsettings = hb_dict_init();
+    hb_value_array_t *avfilters = hb_value_array_init();
+    hb_dict_t *avfilter   = hb_dict_init();
+    hb_dict_t *avsettings = hb_dict_init();
 
     hb_dict_set_string(avsettings, "pix_fmts", format);
     hb_dict_set(avfilter, "format", avsettings);

--- a/libhb/handbrake/common.h
+++ b/libhb/handbrake/common.h
@@ -414,6 +414,7 @@ const char* const* hb_video_encoder_get_presets (int encoder);
 const char* const* hb_video_encoder_get_tunes   (int encoder);
 const char* const* hb_video_encoder_get_profiles(int encoder);
 const char* const* hb_video_encoder_get_levels  (int encoder);
+const int*         hb_video_encoder_get_pix_fmts(int encoder);
 
 void  hb_audio_quality_get_limits(uint32_t codec, float *low, float *high, float *granularity, int *direction);
 float hb_audio_quality_get_best(uint32_t codec, float quality);

--- a/libhb/handbrake/common.h
+++ b/libhb/handbrake/common.h
@@ -409,6 +409,7 @@ const char* hb_video_quality_get_name(uint32_t codec);
 int         hb_video_quality_is_supported(uint32_t codec);
 
 int                hb_video_encoder_is_supported(int encoder);
+int                hb_video_encoder_pix_fmt_is_supported(int encoder, int pix_fmt);
 int                hb_video_encoder_get_depth   (int encoder);
 const char* const* hb_video_encoder_get_presets (int encoder);
 const char* const* hb_video_encoder_get_tunes   (int encoder);
@@ -609,7 +610,10 @@ struct hb_job_s
     char           *encoder_level;
     int             areBframes;
 
-    int             pix_fmt;
+    // Pixel format from decoder to the end of the filters chain
+    int             input_pix_fmt;
+    // Pixel format from the end of filters chain to the encoder
+    int             output_pix_fmt;
     int             color_prim;
     int             color_transfer;
     int             color_matrix;
@@ -1422,6 +1426,7 @@ enum
     HB_FILTER_GRAYSCALE,
     HB_FILTER_PAD,
     HB_FILTER_COLORSPACE,
+    HB_FILTER_FORMAT,
 
     // Finally filters that don't care what order they are in,
     // except that they must be after the above filters

--- a/libhb/handbrake/hbffmpeg.h
+++ b/libhb/handbrake/hbffmpeg.h
@@ -62,5 +62,6 @@ void            hb_avframe_set_video_buffer_flags(hb_buffer_t * buf,
 int hb_av_encoder_present(int encoder);
 const char* const* hb_av_profile_get_names(int encoder);
 const char* const* hb_av_level_get_names(int encoder);
+const int* hb_av_get_pix_fmts(int encoder);
 
 #endif // HANDBRAKE_FFMPEG_H

--- a/libhb/handbrake/internal.h
+++ b/libhb/handbrake/internal.h
@@ -476,6 +476,7 @@ extern hb_filter_object_t hb_filter_unsharp;
 extern hb_filter_object_t hb_filter_avfilter;
 extern hb_filter_object_t hb_filter_mt_frame;
 extern hb_filter_object_t hb_filter_colorspace;
+extern hb_filter_object_t hb_filter_format;
 
 #if HB_PROJECT_FEATURE_QSV
 extern hb_filter_object_t hb_filter_qsv;

--- a/libhb/handbrake/qsv_common.h
+++ b/libhb/handbrake/qsv_common.h
@@ -186,6 +186,7 @@ static const char* const hb_qsv_preset_names2[] = { "speed", "balanced", "qualit
 const char* const* hb_qsv_preset_get_names();
 const char* const* hb_qsv_profile_get_names(int encoder);
 const char* const* hb_qsv_level_get_names(int encoder);
+const int* hb_qsv_get_pix_fmts(int encoder);
 
 const char* hb_qsv_video_quality_get_name(uint32_t codec);
 void hb_qsv_video_quality_get_limits(uint32_t codec, float *low, float *high, float *granularity, int *direction);

--- a/libhb/hb.c
+++ b/libhb/hb.c
@@ -945,7 +945,7 @@ hb_image_t * hb_get_preview3(hb_handle_t * h, int picture,
         ii++;
     }
 
-    job->pix_fmt = init.pix_fmt;
+    job->output_pix_fmt = init.pix_fmt;
     job->color_prim = init.color_prim;
     job->color_transfer = init.color_transfer;
     job->color_matrix = init.color_matrix;

--- a/libhb/hb_json.c
+++ b/libhb/hb_json.c
@@ -620,8 +620,10 @@ hb_dict_t* hb_job_to_dict( const hb_job_t * job )
     hb_dict_set(source_dict, "Range", range_dict);
 
     hb_dict_t *video_dict = hb_dict_get(dict, "Video");
-    hb_dict_set(video_dict, "ColorFormat",
-                hb_value_int(job->pix_fmt));
+    hb_dict_set(video_dict, "ColorInputFormat",
+                hb_value_int(job->input_pix_fmt));
+    hb_dict_set(video_dict, "ColorOutputFormat",
+                hb_value_int(job->output_pix_fmt));
     hb_dict_set(video_dict, "ColorRange",
                 hb_value_int(job->color_range));
     hb_dict_set(video_dict, "ColorPrimaries",
@@ -1015,7 +1017,7 @@ hb_job_t* hb_dict_to_job( hb_handle_t * h, hb_dict_t *dict )
     "s?{s:i, s:i},"
     // Video {Codec, Quality, Bitrate, Preset, Tune, Profile, Level, Options
     //       TwoPass, Turbo,
-    //       ColorFormat, ColorRange,
+    //       ColorInputFormat, ColorOutputFormat, ColorRange,
     //       ColorPrimaries, ColorTransfer, ColorMatrix,
     //       Mastering,
     //       ContentLightLevel,
@@ -1023,7 +1025,7 @@ hb_job_t* hb_dict_to_job( hb_handle_t * h, hb_dict_t *dict )
     //       QSV {Decode, AsyncDepth, AdapterIndex}}
     "s:{s:o, s?F, s?i, s?s, s?s, s?s, s?s, s?s,"
     "   s?b, s?b,"
-    "   s?i, s?i,"
+    "   s?i, s?i, s?i,"
     "   s?i, s?i, s?i,"
     "   s?o,"
     "   s?o,"
@@ -1070,7 +1072,8 @@ hb_job_t* hb_dict_to_job( hb_handle_t * h, hb_dict_t *dict )
             "Options",              unpack_s(&video_options),
             "TwoPass",              unpack_b(&job->twopass),
             "Turbo",                unpack_b(&job->fastfirstpass),
-            "ColorFormat",          unpack_i(&job->pix_fmt),
+            "ColorInputFormat",     unpack_i(&job->input_pix_fmt),
+            "ColorOutputFormat",    unpack_i(&job->output_pix_fmt),
             "ColorRange",           unpack_i(&job->color_range),
             "ColorPrimaries",       unpack_i(&job->color_prim),
             "ColorTransfer",        unpack_i(&job->color_transfer),

--- a/libhb/hbavfilter.c
+++ b/libhb/hbavfilter.c
@@ -99,12 +99,6 @@ hb_avfilter_graph_init(hb_value_t * settings, hb_filter_init_t * init)
         hb_error("hb_avfilter_graph_init: avfilter_graph_alloc failed");
         goto fail;
     }
-#if HB_PROJECT_FEATURE_QSV
-    if (!hb_qsv_hw_filters_are_enabled(graph->job))
-#endif
-    {
-        av_opt_set(graph->avgraph, "scale_sws_opts", "lanczos+accurate_rnd", 0);
-    }
     result = avfilter_graph_parse2(graph->avgraph, settings_str, &in, &out);
     if (result < 0 || in == NULL || out == NULL)
     {
@@ -351,6 +345,7 @@ void hb_avfilter_combine( hb_list_t * list)
             case HB_FILTER_PAD:
             case HB_FILTER_ROTATE:
             case HB_FILTER_COLORSPACE:
+            case HB_FILTER_FORMAT:
             {
                 settings = pv->avfilters;
             } break;

--- a/libhb/qsv_common.c
+++ b/libhb/qsv_common.c
@@ -4236,7 +4236,6 @@ int hb_qsv_sanitize_filter_list(hb_job_t *job)
             }
         }
     }
-
     return 0;
 }
 

--- a/libhb/qsv_common.c
+++ b/libhb/qsv_common.c
@@ -4236,6 +4236,7 @@ int hb_qsv_sanitize_filter_list(hb_job_t *job)
             }
         }
     }
+
     return 0;
 }
 

--- a/libhb/qsv_common.c
+++ b/libhb/qsv_common.c
@@ -1359,7 +1359,7 @@ int hb_qsv_decode_is_enabled(hb_job_t *job)
     return ((job != NULL && job->qsv.decode) &&
             (job->title->video_decode_support & HB_DECODE_SUPPORT_QSV)) &&
             hb_qsv_decode_codec_supported_codec(hb_qsv_get_adapter_index(),
-            job->title->video_codec_param, job->pix_fmt);
+            job->title->video_codec_param, job->input_pix_fmt);
 }
 
 static int hb_dxva2_device_check();
@@ -3612,17 +3612,17 @@ int hb_qsv_attach_surface_to_video_buffer(hb_job_t *job, hb_buffer_t* buf, int i
         DXGI_FORMAT texture_format;
         ID3D11Texture2D* output_surface;
         D3D11_TEXTURE2D_DESC desc = { 0 };
-        if (job->pix_fmt == AV_PIX_FMT_NV12)
+        if (job->input_pix_fmt == AV_PIX_FMT_NV12)
         {
             texture_format = DXGI_FORMAT_NV12;
         }
-        else if(job->pix_fmt == AV_PIX_FMT_P010)
+        else if(job->input_pix_fmt == AV_PIX_FMT_P010)
         {
             texture_format = DXGI_FORMAT_P010;
         }
         else
         {
-            hb_error("hb_qsv_attach_surface_to_video_buffer: unsupported texture_format=%d", job->pix_fmt);
+            hb_error("hb_qsv_attach_surface_to_video_buffer: unsupported texture_format=%d", job->input_pix_fmt);
             return -1;
         }
         hb_qsv_get_free_surface_from_pool_with_range(hb_qsv_frames_ctx, 0, HB_QSV_POOL_SURFACE_SIZE, &mid, &surface);

--- a/libhb/qsv_common.c
+++ b/libhb/qsv_common.c
@@ -158,6 +158,16 @@ static hb_triplet_t hb_qsv_h265_levels[] =
     { NULL,                            },
 };
 
+static const enum AVPixelFormat hb_qsv_pix_fmts[] =
+{
+    AV_PIX_FMT_NV12, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
+static const enum AVPixelFormat hb_qsv_10bit_pix_fmts[] =
+{
+    AV_PIX_FMT_P010LE, AV_PIX_FMT_YUV420P10, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
 #define MFX_IMPL_VIA_MASK(impl) (0x0f00 & (impl))
 
 // check available Intel Media SDK version against a minimum
@@ -2209,6 +2219,21 @@ const char* const* hb_qsv_level_get_names(int encoder)
         default:
             return NULL;
     }
+}
+
+const int* hb_qsv_get_pix_fmts(int encoder)
+{
+    switch (encoder)
+    {
+        case HB_VCODEC_QSV_H264:
+        case HB_VCODEC_QSV_H265:
+            return hb_qsv_pix_fmts;
+        case HB_VCODEC_QSV_H265_10BIT:
+            return hb_qsv_10bit_pix_fmts;
+
+         default:
+             return hb_qsv_pix_fmts;
+     }
 }
 
 const char* hb_qsv_video_quality_get_name(uint32_t codec)

--- a/libhb/qsv_common.c
+++ b/libhb/qsv_common.c
@@ -158,6 +158,7 @@ static hb_triplet_t hb_qsv_h265_levels[] =
     { NULL,                            },
 };
 
+#if defined(_WIN32) || defined(__MINGW32__)
 static const enum AVPixelFormat hb_qsv_pix_fmts[] =
 {
     AV_PIX_FMT_NV12, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
@@ -167,6 +168,17 @@ static const enum AVPixelFormat hb_qsv_10bit_pix_fmts[] =
 {
     AV_PIX_FMT_P010LE, AV_PIX_FMT_YUV420P10, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
 };
+#else
+static const enum AVPixelFormat hb_qsv_pix_fmts[] =
+{
+    AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+
+static const enum AVPixelFormat hb_qsv_10bit_pix_fmts[] =
+{
+    AV_PIX_FMT_YUV420P10, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE
+};
+#endif
 
 #define MFX_IMPL_VIA_MASK(impl) (0x0f00 & (impl))
 

--- a/libhb/sync.c
+++ b/libhb/sync.c
@@ -371,7 +371,7 @@ static hb_buffer_t * CreateBlackBuf( sync_stream_t * stream,
     {
         if (buf == NULL)
         {
-            buf = hb_frame_buffer_init(stream->common->job->pix_fmt,
+            buf = hb_frame_buffer_init(stream->common->job->input_pix_fmt,
                                    stream->common->job->title->geometry.width,
                                    stream->common->job->title->geometry.height);
             uint8_t *planes[4];
@@ -381,7 +381,7 @@ static hb_buffer_t * CreateBlackBuf( sync_stream_t * stream,
                 planes[i] = buf->plane[i].data;
                 linesizes[i] = buf->plane[i].stride;
             }
-            av_image_fill_black(planes, linesizes, stream->common->job->pix_fmt,
+            av_image_fill_black(planes, linesizes, stream->common->job->input_pix_fmt,
                                 stream->common->job->color_range, buf->f.width, buf->f.height);
             buf->f.color_prim = stream->common->job->title->color_prim;
             buf->f.color_transfer = stream->common->job->title->color_transfer;

--- a/libhb/work.c
+++ b/libhb/work.c
@@ -444,13 +444,13 @@ void hb_display_job_info(hb_job_t *job)
 #if HB_PROJECT_FEATURE_QSV
     if (hb_qsv_decode_is_enabled(job))
     {
-        hb_log("   + decoder: %s %d-bit",
-               hb_qsv_decode_get_codec_name(title->video_codec_param), hb_get_bit_depth(job->pix_fmt));
+        hb_log("   + decoder: %s %d-bit (%s)",
+               hb_qsv_decode_get_codec_name(title->video_codec_param), hb_get_bit_depth(job->pix_fmt), av_get_pix_fmt_name(job->pix_fmt));
     }
     else
 #endif
     {
-        hb_log("   + decoder: %s %d-bit", title->video_codec_name, hb_get_bit_depth(job->pix_fmt));
+        hb_log("   + decoder: %s %d-bit (%s)", title->video_codec_name, hb_get_bit_depth(job->pix_fmt), av_get_pix_fmt_name(job->pix_fmt));
     }
 
     if( title->video_bitrate )

--- a/libhb/work.c
+++ b/libhb/work.c
@@ -1345,7 +1345,7 @@ static void sanitize_filter_list(hb_job_t *job, hb_geometry_t src_geo)
         const int *encoder_pix_fmts = hb_video_encoder_get_pix_fmts(job->vcodec);
         int encoder_pix_fmt = *encoder_pix_fmts;
 
-#if defined(_WIN32) || defined(__MINGW32__)
+#if HB_PROJECT_FEATURE_QSV && (defined( _WIN32 ) || defined( __MINGW32__ ))
         if (!hb_qsv_full_path_is_enabled(job))
         {
             // Formats supported by QSV pipeline via system memory

--- a/libhb/work.c
+++ b/libhb/work.c
@@ -445,12 +445,12 @@ void hb_display_job_info(hb_job_t *job)
     if (hb_qsv_decode_is_enabled(job))
     {
         hb_log("   + decoder: %s %d-bit (%s)",
-               hb_qsv_decode_get_codec_name(title->video_codec_param), hb_get_bit_depth(job->pix_fmt), av_get_pix_fmt_name(job->pix_fmt));
+               hb_qsv_decode_get_codec_name(title->video_codec_param), hb_get_bit_depth(job->input_pix_fmt), av_get_pix_fmt_name(job->input_pix_fmt));
     }
     else
 #endif
     {
-        hb_log("   + decoder: %s %d-bit (%s)", title->video_codec_name, hb_get_bit_depth(job->pix_fmt), av_get_pix_fmt_name(job->pix_fmt));
+        hb_log("   + decoder: %s %d-bit (%s)", title->video_codec_name, hb_get_bit_depth(job->input_pix_fmt), av_get_pix_fmt_name(job->input_pix_fmt));
     }
 
     if( title->video_bitrate )
@@ -1271,8 +1271,10 @@ static int sanitize_audio(hb_job_t *job)
     return 0;
 }
 
-static void sanitize_filter_list(hb_list_t *list, hb_geometry_t src_geo)
+static void sanitize_filter_list(hb_job_t *job, hb_geometry_t src_geo)
 {
+    hb_list_t *list = job->list_filter;
+
     // Add selective deinterlacing mode if comb detection is enabled
     if (hb_filter_find(list, HB_FILTER_COMB_DETECT) != NULL)
     {
@@ -1334,6 +1336,15 @@ static void sanitize_filter_list(hb_list_t *list, hb_geometry_t src_geo)
                 hb_log("Skipping crop/scale filter");
             }
         }
+    }
+
+    const int *encoder_pix_fmts = hb_video_encoder_get_pix_fmts(job->vcodec);
+    const int encoder_pix_fmt = *encoder_pix_fmts;
+
+    if (hb_video_encoder_pix_fmt_is_supported(job->vcodec, job->input_pix_fmt) == 0)
+    {
+        hb_filter_object_t *filter = hb_filter_init(HB_FILTER_FORMAT);
+        hb_add_filter(job, filter, hb_strdup_printf("format=%s", av_get_pix_fmt_name(encoder_pix_fmt)));
     }
 }
 
@@ -1400,7 +1411,11 @@ static void do_job(hb_job_t *job)
     {
         hb_filter_init_t init;
 
-        sanitize_filter_list(job->list_filter, title->geometry);
+        // Select the optimal pixel format
+        // for the pipeline
+        job->input_pix_fmt = hb_get_best_pix_fmt(job);
+
+        sanitize_filter_list(job, title->geometry);
 
 #if HB_PROJECT_FEATURE_QSV && (defined( _WIN32 ) || defined( __MINGW32__ ))
         // sanitize_qsv looks for subtitle render filter, so must happen after
@@ -1420,7 +1435,7 @@ static void do_job(hb_job_t *job)
         init.time_base.num = 1;
         init.time_base.den = 90000;
         init.job = job;
-        init.pix_fmt = hb_get_best_pix_fmt(job);
+        init.pix_fmt = job->input_pix_fmt;
         init.color_range = AVCOL_RANGE_MPEG;
 
         init.color_prim = title->color_prim;
@@ -1446,7 +1461,7 @@ static void do_job(hb_job_t *job)
             }
             i++;
         }
-        job->pix_fmt = init.pix_fmt;
+        job->output_pix_fmt = init.pix_fmt;
         job->color_prim = init.color_prim;
         job->color_transfer = init.color_transfer;
         job->color_matrix = init.color_matrix;

--- a/libhb/work.c
+++ b/libhb/work.c
@@ -1338,13 +1338,27 @@ static void sanitize_filter_list(hb_job_t *job, hb_geometry_t src_geo)
         }
     }
 
-    // Some encoders require a specific input pixel format
-    // that could be different from the current pipeline format.
-    const int *encoder_pix_fmts = hb_video_encoder_get_pix_fmts(job->vcodec);
-    const int encoder_pix_fmt = *encoder_pix_fmts;
-
     if (hb_video_encoder_pix_fmt_is_supported(job->vcodec, job->input_pix_fmt) == 0)
     {
+        // Some encoders require a specific input pixel format
+        // that could be different from the current pipeline format.
+        const int *encoder_pix_fmts = hb_video_encoder_get_pix_fmts(job->vcodec);
+        int encoder_pix_fmt = *encoder_pix_fmts;
+
+#if defined(_WIN32) || defined(__MINGW32__)
+        if (!hb_qsv_full_path_is_enabled(job))
+        {
+            // Formats supported by QSV pipeline via system memory
+            if (encoder_pix_fmt != AV_PIX_FMT_YUV420P10 &&
+                encoder_pix_fmt != AV_PIX_FMT_YUV420P)
+            {
+                // TODO: skip the first AV_PIX_FMT_NV12 or AV_PIX_FMT_P010LE formats, prefer second format for system memory
+                encoder_pix_fmts++;
+                encoder_pix_fmt = *encoder_pix_fmts;
+            }
+        }
+#endif
+
         hb_filter_object_t *filter = hb_filter_init(HB_FILTER_FORMAT);
         hb_add_filter(job, filter, hb_strdup_printf("format=%s", av_get_pix_fmt_name(encoder_pix_fmt)));
     }

--- a/libhb/work.c
+++ b/libhb/work.c
@@ -1338,6 +1338,8 @@ static void sanitize_filter_list(hb_job_t *job, hb_geometry_t src_geo)
         }
     }
 
+    // Some encoders require a specific input pixel format
+    // that could be different from the current pipeline format.
     const int *encoder_pix_fmts = hb_video_encoder_get_pix_fmts(job->vcodec);
     const int encoder_pix_fmt = *encoder_pix_fmts;
 

--- a/libhb/work.c
+++ b/libhb/work.c
@@ -1338,6 +1338,29 @@ static void sanitize_filter_list(hb_job_t *job, hb_geometry_t src_geo)
         }
     }
 
+#if HB_PROJECT_FEATURE_QSV && (defined( _WIN32 ) || defined( __MINGW32__ ))
+        // sanitize_qsv looks for subtitle render filter, so must happen after
+        // sanitize_subtitle
+        if (hb_qsv_is_enabled(job))
+        {
+            hb_qsv_sanitize_filter_list(job);
+            // In case of video memory need to set formats after satitize filters 
+            // to ensure that full QSV pipeline supported
+            if (hb_qsv_full_path_is_enabled(job))
+            {
+                // Formats supported in QSV pipeline via video memory
+                if (job->input_pix_fmt == AV_PIX_FMT_YUV420P10)
+                {
+                    job->input_pix_fmt = AV_PIX_FMT_P010LE;
+                }
+                else if (job->input_pix_fmt == AV_PIX_FMT_YUV420P)
+                {
+                    job->input_pix_fmt = AV_PIX_FMT_NV12;
+                }
+            }
+        }
+#endif
+
     if (hb_video_encoder_pix_fmt_is_supported(job->vcodec, job->input_pix_fmt) == 0)
     {
         // Some encoders require a specific input pixel format
@@ -1433,20 +1456,6 @@ static void do_job(hb_job_t *job)
 
         sanitize_filter_list(job, title->geometry);
 
-#if HB_PROJECT_FEATURE_QSV && (defined( _WIN32 ) || defined( __MINGW32__ ))
-        // sanitize_qsv looks for subtitle render filter, so must happen after
-        // sanitize_subtitle
-        if (hb_qsv_is_enabled(job))
-        {
-            result = hb_qsv_sanitize_filter_list(job);
-            if (result)
-            {
-                *job->done_error = HB_ERROR_WRONG_INPUT;
-                *job->die = 1;
-                goto cleanup;
-            }
-        }
-#endif
         memset(&init, 0, sizeof(init));
         init.time_base.num = 1;
         init.time_base.den = 90000;

